src/resolvers/dns.rs

```diff
@@ -214,18 +219,35 @@ impl DnsResolverWorker {
         // whether they responded in time or not.
         let first_result = Arc::new(Mutex::new(None));
 
+        // If we get a `NoRecordsFound` error from a server, we won't consider
+        // the client to have errored, but we also don't want to take that
+        // result as authoritative: if dynamic resolution is in play, we might
+        // have one client that only has records for other DNS servers and other
+        // clients that only have records for backends.
+        //
+        // If at least one client returns `NoRecordsFound` _and_ no clients
+        // return records, we'll return `Some(_)` with an empty list of
+        // backends, to honor what DNS has told us.
+        let saw_no_records_found = Arc::new(AtomicBool::new(false));
```

Comment:
I think this comment is the most important bit - we don't want to treat `NoRecordsFound` as a client _error_, but we also don't want to take it as the first result (particularly due to dynamic resolution).